### PR TITLE
Add gzip compression option for the JWT Claims part

### PR DESCRIPTION
**Pull request by szank**

### CODE REVIEW

**Comment by bbrks** [comment]
:+1: for compression support, but I'm reluctant to support a breaking change to token.New()

Maybe a token.NewCompressed() or something? And let the old one internally default to jwt.CompressionNone

**Comment by elithrar** [comment]
Do any other major JWT libs do gzip compression of claims? (out of
curiosity!)

**Comment by bbrks** [comment]
[jose-jwt](https://github.com/dvsekhvalnov/jose-jwt) and its various ports ([jose2go](https://github.com/dvsekhvalnov/jose2go), [jose.4.j](https://bitbucket.org/b_c/jose4j/wiki/Home)) support deflate out of the box. But not gzip.

**Comment by elithrar** [comment]
I'd just be cautious about actually gzipping claims—the spec doesn't say
you can't, but the spec is reasonably prescriptive and doesn't say you can,
either ;)

Importantly though, what browser-based gzip libs exist for decompressing
the claims, and are those performant on mobile?
https://github.com/nodeca/pako looks solid but focuses on node
benchmarks/examples.

**Comment by bbrks** [comment]
To be fair, you could just compress any huge byte slices before you put it in the JWT. And decompress when you want to use the data. This has the advantage of sticking to the spec whilst giving you a huge compressed blob.

screen shot 2015-12-15 at 10 40 18
screen shot 2015-12-15 at 10 41 06

**Comment by szank** [comment]
First, I agree, that adding a new method and not breaking the existing API could be better. I wanted to see what @dgrijalva  would say about it, but it seems that he is taking a break from github right now :). 

Also, gzipping the claims is not enforced. We use JWT in a RESTful application, the claims won't be read by the client, only by the server which have created them in the first place.  It is useful only in such restricted use case, but we are not trying to make everyone happy here anyway. 

Also, yes, I can gzip my data before passing it to JWT, but I try to contribute some useful changes to the upstream if I can. I rather have it done in JWT, makes the code cleaner and more reusable.

**Comment by dgrijalva** [comment]
Why not just compress the whole blob?

**Comment by dgrijalva** [comment]
It seems the right place to add a feature like this would be in the spec.  Though, there are several other ways to implement this without violating the spec.

* If all clients are going to treat the token as an opaque string, you could just compress the whole token and stick that in the header.
* For a more forward looking approach, SPDY (HTTP/2) support full header compression, making this moot.

For this library, I'd be most comfortable with making compression possible with an external add on.  I'd like to adhere to the spec as much as possible, without closing doors for users to do what they'd like.

**Comment by kofalt** [comment]
Possibly dumb question: Won't compression at the transfer layer take care of this? Why bother?

File: token.go [hunk 1]
```diff
@@ -61,22 +62,27 @@ func (t *Token) SignedString(key interface{}) (string, error) {
...
+		return "", err
+	}
+	parts = append(parts, EncodeSegment(jsonValue))
 
-		parts[i] = EncodeSegment(jsonValue)
+	if jsonValue, err = json.Marshal(t.Claims); err != nil {
+		return "", err
 	}
+
+	if compression, err = getCompressionMethod(t.Header["cpr"]); err != nil {
```
Comment (AlmogBaku) [inline token.go]: this may break older versions of the token

**Comment by AlmogBaku** [comment]
@dgrijalva how can you make this as an addon? by implementing a wrapper encryption method?

**Comment by dgrijalva** [comment]
Here's a function that will compress and re-base64url encode the content.  Because of the double encoding, the content ends up being larger.  This may not hold true for larger tokens.  Give it a try and let me know what you think?

```go
func compressedToken(s string)string {
	buf := new(bytes.Buffer)
	enc := base64.NewEncoder(base64.URLEncoding, buf)
	wr := gzip.NewWriter(enc)
	wr.Write([]byte(s))
	wr.Close()
	enc.Close()

	return string(buf.Bytes())
}
```

**Comment by dgrijalva** [comment]
Unless this gets added to the spec, I don't think its inclusion is wise.  If you need new hooks allow you to accomplish this outside the library, let's have that discussion instead.

**Comment by Ocramius** [comment]
Just poking back here: regardless of the spec, we came to the conclusion that most mediums for storage or transfer of JWT tokens are already supporting their lower-layer compression (before filesystem or network interaction).

Ref: https://github.com/lcobucci/jwt/issues/82#issuecomment-209516213 (just to provide arguments, should this come up again)

**Comment by asambeka** [comment]
Compression is part of JOSE specs on IANA, reference: https://www.iana.org/assignments/jose/jose.xhtml and also RFC 7516.
Header parameter: zip
Supported values : DEF (Deflate from RFC1951)

**Pull request closed without merging**